File: elixir/lib/nucleus/models/pubsub.ex
defmodule Nucleus.Models.PubSub do
  @moduledoc """
  PubSub model — PUBSUB_PUBLISH, PUBSUB_CHANNELS, PUBSUB_SUBSCRIBERS.

  ## Example

      alias Nucleus.Models.PubSub

      {:ok, count} = PubSub.publish(client, "notifications", "Hello!")
      {:ok, channels} = PubSub.channels(client)
      {:ok, subs} = PubSub.subscribers(client, "notifications")
  """

  @type client :: Nucleus.Client.t()

  @doc "Publishes a message to a channel. Returns the number of subscribers reached."
  @spec publish(client(), String.t(), String.t()) :: {:ok, integer()} | {:error, term()}
  def publish(client, channel, message) do
    with :ok <- Nucleus.Client.require_nucleus(client, "PubSub.publish") do
      case Nucleus.Client.query(client, "SELECT PUBSUB_PUBLISH($1, $2)", [channel, message]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end

  @doc "Lists active channels, optionally filtered by pattern."
  @spec channels(client(), String.t() | nil) :: {:ok, [String.t()]} | {:error, term()}
  def channels(client, pattern \\ nil) do
    with :ok <- Nucleus.Client.require_nucleus(client, "PubSub.channels") do
      result =
        if pattern do
          Nucleus.Client.query(client, "SELECT PUBSUB_CHANNELS($1)", [pattern])
        else
          Nucleus.Client.query(client, "SELECT PUBSUB_CHANNELS()", [])
        end

      case result do
        {:ok, %{rows: [[raw]]}} when is_binary(raw) and raw != "" ->
          {:ok, String.split(raw, ",")}

        {:ok, _} ->
          {:ok, []}

        {:error, _} = error ->
          error
      end
    end
  end

  @doc "Returns the number of subscribers for a channel."
  @spec subscribers(client(), String.t()) :: {:ok, integer()} | {:error, term()}
  def subscribers(client, channel) do
    with :ok <- Nucleus.Client.require_nucleus(client, "PubSub.subscribers") do
      case Nucleus.Client.query(client, "SELECT PUBSUB_SUBSCRIBERS($1)", [channel]) do
        {:ok, %{rows: [[val]]}} -> {:ok, val}
        {:error, _} = error -> error
      end
    end
  end
end

File: rs/crates/neutron-nucleus/src/models/pubsub.rs
//! PubSub model — PUBSUB_PUBLISH, PUBSUB_CHANNELS, PUBSUB_SUBSCRIBERS.

use crate::error::NucleusError;
use crate::pool::NucleusPool;

/// Handle for PubSub operations.
pub struct PubSubModel {
    pool: NucleusPool,
}

impl PubSubModel {
    pub(crate) fn new(pool: NucleusPool) -> Self {
        Self { pool }
    }

    /// Publish a message on a channel. Returns the number of subscribers reached.
    pub async fn publish(&self, channel: &str, message: &str) -> Result<i64, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT PUBSUB_PUBLISH($1, $2)", &[&channel, &message])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, i64>(0))
    }

    /// Return active PubSub channels matching an optional pattern.
    pub async fn channels(&self, pattern: Option<&str>) -> Result<Vec<String>, NucleusError> {
        let conn = self.pool.get().await?;
        let row = if let Some(pat) = pattern {
            conn.client()
                .query_one("SELECT PUBSUB_CHANNELS($1)", &[&pat])
                .await
                .map_err(NucleusError::Query)?
        } else {
            conn.client()
                .query_one("SELECT PUBSUB_CHANNELS()", &[])
                .await
                .map_err(NucleusError::Query)?
        };
        let raw: String = row.get(0);
        if raw.is_empty() {
            return Ok(Vec::new());
        }
        Ok(raw.split(',').map(|s| s.trim().to_string()).collect())
    }

    /// Return the number of subscribers on a channel.
    pub async fn subscribers(&self, channel: &str) -> Result<i64, NucleusError> {
        let conn = self.pool.get().await?;
        let row = conn
            .client()
            .query_one("SELECT PUBSUB_SUBSCRIBERS($1)", &[&channel])
            .await
            .map_err(NucleusError::Query)?;
        Ok(row.get::<_, i64>(0))
    }
}
